File: Week 23/cpp/1.AdvancedDP.cpp
/*
 * WEEK 23 - C++ ADVANCED DSA
 * Topic: Advanced Dynamic Programming (Bitmask, Tree, Digit DP, LIS)
 * File: 1.AdvancedDP.cpp
 *
 * CONCEPT:
 *   Bitmask DP encodes a subset as an integer (n <= ~20).
 *   DP on Trees uses post-order recursion.
 *   Digit DP iterates over digit positions with a "tight" flag.
 *   LIS via patience sorting reaches O(n log n).
 *
 * KEY POINTS:
 *   - For TSP, dp[mask][i] = min cost path covering 'mask' ending at i.
 *   - For tree max-indep-set, dp[u][0/1] track "u not taken" / "u taken".
 *   - Digit DP memoises (pos, tight) when "started"; otherwise recompute.
 *   - LIS uses std::lower_bound on a tails vector.
 *
 * ALGORITHM / APPROACH:
 *   See sibling files; identical recurrences.
 *
 * C++-SPECIFIC NOTES vs JAVA:
 *   - Use INT_MAX/2 sentinel; or explicit large constant.
 *   - std::vector<std::vector<int>> for the dp table.
 *   - Use std::lower_bound for LIS binary search.
 *   - Recursion is fine; switch to iterative DFS for very deep trees.
 *
 * DRY RUN:
 *   TSP for 4-city symmetric matrix in main(): optimal = 80
 *   (route 0 -> 1 -> 3 -> 2 -> 0).
 *   count_no_four(10) -> 9 (skip 4).
 *   LIS([10,9,2,5,3,7,101,18]) -> 4.
 *
 * COMPLEXITY:
 *   TSP:      O(n^2 * 2^n) time, O(n * 2^n) space.
 *   Tree DP:  O(n).
 *   Digit DP: O(D * 10) per state.
 *   LIS:      O(n log n).
 */

#include <iostream>
#include <vector>
#include <queue>
#include <unordered_map>
#include <algorithm>
#include <climits>
#include <string>
#include <array>

constexpr int INF_HALF = INT_MAX / 2;

int tsp_min_cost(const std::vector<std::vector<int>>& dist) {
    int n = dist.size();
    int FULL = (1 << n) - 1;
    std::vector<std::vector<int>> dp(1 << n, std::vector<int>(n, INF_HALF));
    dp[1][0] = 0;
    for (int mask = 1; mask <= FULL; ++mask) {
        for (int u = 0; u < n; ++u) {
            if (!(mask & (1 << u)) || dp[mask][u] == INF_HALF) continue;
            for (int v = 0; v < n; ++v) {
                if (mask & (1 << v)) continue;
                int nm = mask | (1 << v);
                if (dp[mask][u] + dist[u][v] < dp[nm][v])
                    dp[nm][v] = dp[mask][u] + dist[u][v];
            }
        }
    }
    int ans = INT_MAX;
    for (int u = 1; u < n; ++u)
        if (dp[FULL][u] != INF_HALF)
            ans = std::min(ans, dp[FULL][u] + dist[u][0]);
    return ans;
}

void dfs_tree(int u, const std::vector<int>& weight,
              const std::vector<std::vector<int>>& children,
              std::vector<std::array<int,2>>& dp,
              std::vector<bool>& visited) {
    visited[u] = true;
    dp[u][1] = weight[u];
    dp[u][0] = 0;
    for (int c : children[u]) {
        if (!visited[c]) {
            dfs_tree(c, weight, children, dp, visited);
            dp[u][1] += dp[c][0];
            dp[u][0] += std::max(dp[c][0], dp[c][1]);
        }
    }
}

int max_indep_set(const std::vector<int>& weight, const std::vector<std::vector<int>>& children) {
    int n = weight.size();
    std::vector<std::array<int,2>> dp(n, {0, 0});
    std::vector<bool> visited(n, false);
    dfs_tree(0, weight, children, dp, visited);
    return std::max(dp[0][0], dp[0][1]);
}

// Digit DP: count numbers in [1..N] without digit '4'
static std::vector<std::vector<int>> digitMemo;
static std::string g_num;

int digit_dp(int pos, int tight, int started) {
    int n = (int)g_num.size();
    if (pos == n) return started ? 1 : 0;
    if (!tight && started && digitMemo[pos][0] != -1) return digitMemo[pos][0];
    int limit = tight ? (g_num[pos] - '0') : 9;
    int count = 0;
    for (int d = 0; d <= limit; ++d) {
        if (d == 4) continue;
        count += digit_dp(pos + 1, tight && (d == limit), started || (d != 0));
    }
    if (!tight && started) digitMemo[pos][0] = count;
    return count;
}

int count_no_four(int N) {
    g_num = std::to_string(N);
    digitMemo.assign(g_num.size(), std::vector<int>(2, -1));
    return digit_dp(0, 1, 0);
}

// LIS in O(n log n) via patience sorting
int lis_optimal(const std::vector<int>& nums) {
    std::vector<int> tails;
    for (int x : nums) {
        auto it = std::lower_bound(tails.begin(), tails.end(), x);
        if (it == tails.end()) tails.push_back(x);
        else *it = x;
    }
    return (int)tails.size();
}

int main() {
    std::cout << "=== TSP with Bitmask DP ===\n";
    std::vector<std::vector<int>> dist4 = {
        {0,10,15,20},{10,0,35,25},{15,35,0,30},{20,25,30,0}
    };
    std::cout << "TSP (4 cities): " << tsp_min_cost(dist4) << "\n";

    std::cout << "\n=== Max Independent Set on Tree ===\n";
    std::vector<int> weights = {1, 2, 3, 4, 5};
    std::vector<std::vector<int>> children = {{1,2},{3,4},{},{},{}};
    std::cout << "Max weight independent set: " << max_indep_set(weights, children) << "\n";

    std::cout << "\n=== Digit DP: Count numbers without digit 4 ===\n";
    std::cout << "Count in [1..10]:  " << count_no_four(10)  << "\n";
    std::cout << "Count in [1..40]:  " << count_no_four(40)  << "\n";
    std::cout << "Count in [1..100]: " << count_no_four(100) << "\n";

    std::cout << "\n=== LIS O(n log n) ===\n";
    std::vector<std::vector<int>> tests = {
        {10,9,2,5,3,7,101,18}, {0,1,0,3,2,3}, {7,7,7,7,7}
    };
    for (auto& t : tests) {
        std::cout << "[ ";
        for (int v : t) std::cout << v << " ";
        std::cout << "] -> LIS = " << lis_optimal(t) << "\n";
    }
    return 0;
}

File: Week 23/python/1.AdvancedDP.py
"""
WEEK 23 - PYTHON ADVANCED DSA
Topic: Advanced Dynamic Programming (Bitmask, DP on Trees, Digit DP, LIS)
File: 1.AdvancedDP.py

CONCEPT:
    Three families of harder DP, plus an O(n log n) LIS.
        Bitmask DP:  state encodes a subset of n items as an integer.
        DP on Trees: post-order recursion; subproblem = subtree of node.
        Digit DP:    count numbers in [1..N] with property P, building
                     digit by digit while tracking a "tight" flag.

KEY POINTS:
    - Bitmask DP works when n <= ~20 because 2^n states are tractable.
    - For tree DP, dp[u][0/1] often distinguishes "u not taken / taken".
    - Digit DP state = (pos, tight, started, ...property-specific flags).
    - LIS via patience sorting: maintain `tails[i]` = smallest end value of
      any increasing subsequence of length i+1; size grows monotonically.

ALGORITHM / APPROACH:
    TSP (bitmask DP):
        dp[mask][i] = min cost path visiting exactly cities in mask, ending at i
        dp[1][0] = 0; for each mask, for each u in mask, for each v not in
                                 mask: dp[mask|1<<v][v] = min(..., dp[mask][u] + dist[u][v])
        answer = min over u: dp[FULL][u] + dist[u][0]

    Max Independent Set (Tree):
        post-order; dp[u][1] = w[u] + sum(dp[c][0]); dp[u][0] = sum(max(dp[c][0],dp[c][1]))

    Digit DP (count numbers 1..N without digit '4'):
        recursion on (pos, tight, started); memo on (pos, tight) when not started

    LIS in O(n log n):
        for x in nums: bisect_left(tails, x); replace tails[idx] with x;
        if idx == len(tails): tails.append(x).

PYTHON-SPECIFIC NOTES vs JAVA:
    - Use functools.lru_cache for memoised digit DP — cleaner than manual table.
    - bisect.bisect_left implements the LIS binary search.
    - Avoid recursion-depth issues by raising sys.setrecursionlimit if needed.

DRY RUN:
    TSP on 4 cities with the matrix in main(): optimal cost = 80.
    Tree weights [1,2,3,4,5], children {0:[1,2],1:[3,4]}:
        dp[2]=[0,3], dp[3]=[0,4], dp[4]=[0,5]
        dp[1] = [0+max(0,4)+max(0,5), 2+0+0] = [9, 2]
        dp[0] = [max(9,2)+max(0,3), 1 + min taken of (1) + min taken of (2)]
              = [9+3, 1 + 9_no? ] ... see code for exact reproduction
    Digit DP: count_no_four(10) = 9 (skip 4); count_no_four(40) skips 4,14,24,34,40 -> 35.
    LIS([10,9,2,5,3,7,101,18]) -> 4 (e.g., 2,3,7,101).

COMPLEXITY:
    TSP:      O(n^2 * 2^n)
    Tree DP:  O(n)
    Digit DP: O(D * 10) per state; states = O(D * states_of_extra_flags)
    LIS:      O(n log n)
"""

from __future__ import annotations
from typing import List, Tuple
from functools import lru_cache
import bisect


# ---------- TSP via Bitmask DP ----------
def tsp_min_cost(dist: List[List[int]]) -> int:
    n = len(dist)
    FULL = (1 << n) - 1
    INF = 10 ** 18
    dp = [[INF] * n for _ in range(1 << n)]
    dp[1][0] = 0  # start at city 0

    for mask in range(1, 1 << n):
        for u in range(n):
            if not (mask & (1 << u)) or dp[mask][u] == INF:
                continue
            for v in range(n):
                if mask & (1 << v):
                    continue
                new_mask = mask | (1 << v)
                cand = dp[mask][u] + dist[u][v]
                if cand < dp[new_mask][v]:
                    dp[new_mask][v] = cand

    ans = INF
    for u in range(1, n):
        if dp[FULL][u] + dist[u][0] < ans:
            ans = dp[FULL][u] + dist[u][0]
    return ans


# ---------- Maximum Independent Set on Tree ----------
def max_indep_set(weight: List[int], children: List[List[int]]) -> int:
    n = len(weight)
    dp = [[0, 0] for _ in range(n)]
    visited = [False] * n

    def dfs(u: int) -> None:
        visited[u] = True
        dp[u][1] = weight[u]
        dp[u][0] = 0
        for c in children[u]:
            if not visited[c]:
                dfs(c)
                dp[u][1] += dp[c][0]
                dp[u][0] += max(dp[c][0], dp[c][1])

    dfs(0)
    return max(dp[0][0], dp[0][1])


# ---------- Digit DP: count integers in [1..N] without digit '4' ----------
def count_no_four(N: int) -> int:
    digits = list(map(int, str(N)))
    L = len(digits)

    @lru_cache(maxsize=None)
    def rec(pos: int, tight: bool, started: bool) -> int:
        if pos == L:
            return 1 if started else 0
        limit = digits[pos] if tight else 9
        total = 0
        for d in range(0, limit + 1):
            if d == 4:
                continue
            total += rec(pos + 1, tight and (d == limit), started or (d != 0))
        return total

    return rec(0, True, False)


# ---------- Longest Increasing Subsequence (O(n log n)) ----------
def lis_optimal(nums: List[int]) -> int:
    tails: List[int] = []
    for x in nums:
        i = bisect.bisect_left(tails, x)
        if i == len(tails):
            tails.append(x)
        else:
            tails[i] = x
    return len(tails)


def main() -> None:
    print("=== TSP with Bitmask DP ===")
    dist4 = [
        [0, 10, 15, 20],
        [10, 0, 35, 25],
        [15, 35, 0, 30],
        [20, 25, 30, 0],
    ]
    print("TSP (4 cities):", tsp_min_cost(dist4))  # 80

    print("\n=== Max Independent Set on Tree ===")
    weights = [1, 2, 3, 4, 5]
    children = [[1, 2], [3, 4], [], [], []]
    print("Max weight independent set:", max_indep_set(weights, children))

    print("\n=== Digit DP: Count numbers without digit 4 ===")
    print("Count in [1..10]: ", count_no_four(10))
    print("Count in [1..40]: ", count_no_four(40))
    print("Count in [1..100]:", count_no_four(100))

    print("\n=== LIS O(n log n) ===")
    for t in ([10, 9, 2, 5, 3, 7, 101, 18], [0, 1, 0, 3, 2, 3], [7, 7, 7, 7, 7]):
        print(t, "-> LIS =", lis_optimal(t))


if __name__ == "__main__":
    main()
